src/util/chooseRenormalize.ts

```diff
@@ -0,0 +1,64 @@
+function mean(values: number[]): number {
+  return values.reduce((memo: number, i: number) => memo + i) / values.length;
+}
+
+/**
+ * This defines a random number between 0 and "mass", which should cover the entire
+ * domain of the probability distribution. Then we will "walk" the probability mass
+ * until we exceed the generated value
+ */
+function basicChooseOne(probs: number[], mass: number): number {
+  const r = mass * Math.random();
+  let i = 0;
+  let s = probs[0];
+  while (s < r) {
+    i += 1;
+    s += probs[i];
+  }
+  return i;
+}
+
+function chooseRenormalize(probs: number[], k: number): number[] {
+  //  If we have more work slots than candidates in the probability distribution,
+  //  we return them all
+  if (k >= probs.length)
+    return Array(probs.length)
+      .fill(null)
+      .map((x: null, index: number) => index);
+
+  // Since we are mutating the probabilities vector, we copy it first
+  const p = probs.map(x => x);
+
+  // This assumes that the probability vector is normalized. If it is not, it
+  // should be changed to the sum of the weights
+  let remaining = 1;
+
```

Comment:
Are we actually making sure that the weights are normalized somewhere?

Reply:
I will sum the probability values.